### PR TITLE
Authentication - refresh token and authorization code flows

### PR DESCRIPTION
**Pull request by geasi**

### CODE REVIEW

File: src/Pipedrive.net/Models/Request/OAuthRefreshTokenRequest.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+            ClientId = clientId;
+            ClientSecret = clientSecret;
+            RefreshToken = refreshToken;
+        }
+
+        [Parameter(Key = "refresh_token")]
+        public string RefreshToken { get; }
+
+        [Parameter(Key = "grant_type")]
+        public string GrantType => "refresh_token";
```
Comment (geasi) [inline src/Pipedrive.net/Models/Request/OAuthRefreshTokenRequest.cs]: @DavidRouyer 
What do you think of this approach?
The grant type could not be another but refresh_token in this scenario.

**Comment by DavidRouyer** [comment]
Can you resolve the conflicts please?

**Review by DavidRouyer — CHANGES_REQUESTED** [review]
Can you resolve the conflicts please?

**Comment by geasi** [comment]
Sorry @DavidRouyer, I was out these days.
Yes, I will take a look and solve it not later than tomorrow.

Thanks

**Comment by geasi** [comment]
Hi @DavidRouyer ,

I finally could review the conflicts and resolved the issues, ok?
Thanks!

File: src/Pipedrive.net/Models/Request/OAuthRefreshTokenRequest.cs [hunk 2]
```diff
@@ -21,11 +21,11 @@ public OAuthRefreshTokenRequest(string clientId, string clientSecret, string ref
         {
             Ensure.ArgumentNotNullOrEmptyString(clientId, nameof(clientId));
             Ensure.ArgumentNotNullOrEmptyString(clientSecret, nameof(clientSecret));
+            Ensure.ArgumentNotNullOrEmptyString(refreshToken, nameof(refreshToken));
 
             ClientId = clientId;
             ClientSecret = clientSecret;
             RefreshToken = refreshToken;
-            GrantType = "refresh_token";
```
Comment (DavidRouyer) [inline src/Pipedrive.net/Models/Request/OAuthRefreshTokenRequest.cs]: Don't remove the GrantType

File: src/Pipedrive.net/Clients/OAuthClient.cs [hunk 3]
```diff
@@ -83,5 +83,40 @@ public async Task<OAuthToken> RefreshAccessToken(OAuthRefreshTokenRequest reques
             var response = await connection.Post<OAuthToken>(endPoint, body, "application/json", null, hostAddress).ConfigureAwait(false);
             return response.Body;
         }
+        
+        public async Task<OAuthToken> RefreshAccessToken(OAuthRefreshTokenRequest request)
```
Comment (DavidRouyer) [inline src/Pipedrive.net/Clients/OAuthClient.cs]: Duplicated `RefreshAccessToken`: it already exists

File: src/Pipedrive.net/Clients/OAuthClient.cs [hunk 4]
```diff
@@ -83,5 +83,40 @@ public async Task<OAuthToken> RefreshAccessToken(OAuthRefreshTokenRequest reques
...
+
+            var body = new FormUrlEncodedContent(parameters);
+
+            connection.Credentials = new Credentials(clientId, clientSecret, AuthenticationType.Basic);
+
+            var response = await connection.Post<OAuthToken>(endPoint, body, "application/json", null, hostAddress).ConfigureAwait(false);
+            return response.Body;
+        }
+
+        public async Task RevokeToken(OAuthRevokeTokenRequest request)
```
Comment (DavidRouyer) [inline src/Pipedrive.net/Clients/OAuthClient.cs]: Can you point me to the documentation for revoking the token?

**Review by DavidRouyer — CHANGES_REQUESTED** [review]

File: src/Pipedrive.net/Clients/OAuthClient.cs (hunk 4 above)
Comment (DavidRouyer) [reply src/Pipedrive.net/Clients/OAuthClient.cs]: Ok, found it https://pipedrive.readme.io/docs/app-uninstallation#section-token-revocation

File: src/Pipedrive.net/Clients/OAuthClient.cs (hunk 3 above)
Comment (DavidRouyer) [reply src/Pipedrive.net/Clients/OAuthClient.cs]: Ok, found it https://pipedrive.readme.io/docs/app-uninstallation#section-token-revocation

File: src/Pipedrive.net/Models/Request/OAuthRefreshTokenRequest.cs (hunk 2 above)
Comment (geasi) [reply src/Pipedrive.net/Models/Request/OAuthRefreshTokenRequest.cs]: Oww sorry! Fixing this!

File: src/Pipedrive.net/Clients/OAuthClient.cs (hunk 3 above)
Comment (geasi) [reply src/Pipedrive.net/Clients/OAuthClient.cs]: Fixed!

**Comment by geasi** [comment]
@DavidRouyer,
Thank you a lot for reviewing this, I really appreciate it.
Today I had some time to update my local branch and I checked your changes on JsonHttpPipeline, thank you a lot for that. 
I found some methods that stopped working because now we have to specify when the response is a JsonResponse or not.
For example, the method above:
`
        public Task<Organization> Create(NewOrganization data)
        {
            Ensure.ArgumentNotNull(data, nameof(data));

            return ApiConnection.Post<Organization>(ApiUrls.Organizations(), data);
        }
`
Now is not working because the response is not only Organization, it is a JsonResponse<Organization>.
I'm working on a solution for this ok?

Thanks!

**Comment by geasi** [comment]
David,

After reviewing the code for a while I could realize that now, in the ApiConnection class, we have methods that read a JsonResponse result, and methods that don't.
For example the following method:
  public async Task<T> Get<T>(Uri uri, IDictionary<string, string> parameters)
This method has a body that reads the response as a JsonResponse of T.
In the other hand for example, this method, in the same class:
  public async Task<T> Post<T>(Uri uri)
Has a body that does not read the response as a JsonResponse of T, it read only response of T.
The post methods are able to read the response from the token endpoints but not from the create organization endpoint for example.
My suggestion: why don't we keep the same behaviour between the methods in the ApiConnection class reading always from T and introduce the logic of reading (or don't) from a JsonResponse object in the XYZClient classes?
A change I made locally and worked consistently was the above method from OrganizationsClient class:

`public async Task<Organization> Create(NewOrganization data)
        {
            Ensure.ArgumentNotNull(data, nameof(data));

            var result = await ApiConnection.Post<JsonResponse<Organization>>(ApiUrls.Organizations(), data);
            return result.Data;
        }`

**Comment by geasi** [comment]
David,

I just updated the PR with not the full refactoring, but only the methods I could test for now, ok?
With the changes in the Create and Edit methods I made, now the client is able to read the response again, ok?
I would appreciate if you could kindly review it and share your thoughts with me.

Thanks!

**Comment by DavidRouyer** [comment]
@geasi Please revert your last commit so I can merge your PR. If we have to change the JSON responses to POST, PUT and DELETE, I have to edit them for all requests (so, in the `ApiConnection` class). Let me do it.

**Comment by geasi** [comment]
@DavidRouyer I reverted the last commit, ok?
Thanks

**Comment by geasi** [comment]
> @geasi Please revert your last commit so I can merge your PR. If we have to change the JSON responses to POST, PUT and DELETE, I have to edit them for all requests (so, in the `ApiConnection` class). Let me do it.

@DavidRouyer now I understood what you mean. I did not realize that we had two classes now (ApiConnection and Connection) and OAuthClient uses the Connection class and all the other classes are using the ApiConnection, now I got it.
I will make this change on ApiConenction locally and make more tests today.

Thanks a lot, man!

**Review by DavidRouyer — APPROVED** [review]

**Comment by DavidRouyer** [comment]
Thank you for your contribution!

**Pull request merged**